### PR TITLE
Handle newlines in user pills

### PR DESCRIPTION
**Pull request by Johennes**
Fixes: vector-im/element-web#10994

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Handle newlines in user pills ([\#11166](https://github.com/matrix-org/matrix-react-sdk/pull/11166)). Fixes vector-im/element-web#10994.

### CODE REVIEW

File: test/editor/serialize-test.ts [hunk 1]
```diff
@@ -63,6 +63,12 @@ describe("editor/serialize", function () {
             const html = htmlSerializeIfNeeded(model, {});
             expect(html).toBe('<a href="https://matrix.to/#/@user:server">Displayname]</a>');
         });
+        it("displaynames containing a newline work", function () {
+            const pc = createPartCreator();
+            const model = new EditorModel([pc.userPill("Display\nname", "@user:server")], pc);
+            const html = htmlSerializeIfNeeded(model, {});
+            expect(html).toBe('<a href="https://matrix.to/#/@user:server">Display<br>name</a>');
```
[the remark this answers is not in the document]
Comment (richvdh) [reply test/editor/serialize-test.ts:70]: why are we trying to use `<br/>` in the first place? Shouldn't it be `<br>` in modern html?

https://developer.mozilla.org/en-US/docs/Glossary/Void_element#self-closing_tags

File: src/editor/serialize.ts [hunk 2]
```diff
@@ -48,7 +48,9 @@ export function mdSerialize(model: EditorModel): string {
             case Type.UserPill:
                 return (
                     html +
-                    `[${part.text.replace(/[[\\\]]/g, (c) => "\\" + c)}](${makeGenericPermalink(part.resourceId)})`
+                    `[${part.text.replace(/[[\\\]]/g, (c) => "\\" + c).replace(/\n/g, "<br/>")}](${makeGenericPermalink(
+                        part.resourceId,
+                    )})`
```
Comment (richvdh) [inline src/editor/serialize.ts:53]: this is getting a bit hard to read. Can we (a) factor out the part generating the `[...]` content to a local variable; (b) get a comment explaining what it does?

I think it's escaping `[`, `\` and `]` with a `\`, and replacing line break characters with `<br/>` ?

File: test/editor/serialize-test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Johennes) [reply test/editor/serialize-test.ts:70]: My bad, you're right.

File: src/Markdown.ts [hunk 3]
```diff
@@ -22,7 +22,7 @@ import { logger } from "matrix-js-sdk/src/logger";
 
 import { linkify } from "./linkify-matrix";
 
-const ALLOWED_HTML_TAGS = ["sub", "sup", "del", "u"];
+const ALLOWED_HTML_TAGS = ["sub", "sup", "del", "u", "br", "br/"];
```
Comment (richvdh) [inline src/Markdown.ts:25]: presumably `br/` is now redundant here?

File: test/editor/deserialize-test.ts [hunk 4]
```diff
@@ -183,6 +183,14 @@ describe("editor/deserialize", function () {
             expect(parts[1]).toStrictEqual({ type: "user-pill", text: "Alice]", resourceId: "@alice:hs.tld" });
             expect(parts[2]).toStrictEqual({ type: "plain", text: "!" });
         });
+        it("user pill with displayname containing linebreak square bracket", function () {
+            const html = 'Hi <a href="https://matrix.to/#/@alice:hs.tld">Alice<br/>123</a>!';
```
Comment (richvdh) [inline test/editor/deserialize-test.ts:187]: there is no square bracket in this input?

File: src/Markdown.ts (hunk 3 above)
Comment (Johennes) [reply src/Markdown.ts:25]: No, I'd say not. This is actually a bit broader than what we need. The tags listed here are allowed _anywhere_ in a message, not just inside user pills. So if we remove `br/`, user pills would still work but you could not use `<br/>` in a message even though you can us `<br>` which feels a bit non-intuitive to me.

File: test/editor/deserialize-test.ts (hunk 4 above)
Comment (Johennes) [reply test/editor/deserialize-test.ts]: Copy/paste error, apologies. 🙄

**Review by richvdh — APPROVED** [review]
lgtm!

**Pull request merged**